Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel $\rightarrow$ Restart) and then **run all cells** (in the menubar, select Cell $\rightarrow$ Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [1]:
NAME = "Vishrut Vanga"
COLLABORATORS = ""

---

# **CS 210**: Assignment 2 [100 points]


For this assignment, you will write a Python module called **A2sub.py** in which you will define the five functions below. 

We are providing you with a stub file **A2sub.py** that contains the function definitions. **Do not change the definitions.** Functions are defined with hints (noted ->) which specify what types the function is expected to output. You can define your own helper functions, as long as the main function for each question follows the given template.

You should submit the Assignment via CodeBench. Click submit next to the assignment. This will submit a snapshot of your whole directory. We will only grade the file **A2sub.py**

You can submit as many times as you want before the deadline, but each submission creates a new snapshot and **erases the previous submission**.

# Dataset
For this assignment, you will use the dataset located in CodeBench in `../../shared/Datasets/Nov23_East_West_Flights.csv`

The data consists of flight records taken in November 2023 and was retrieved from the Bureau of Transportation Statistics (https://www.transtats.bts.gov/DL_SelectFields.aspx?gnoyr_VQ=FGK&QO_fu146_anzr=b0-gvzr)

You can see the description of the record fields in `../../shared/Datasets/FlightDataRecordDescription.html`

Note that the data you are given is a subset of the original data, that only includes some of the record fields, and is a selection of all cross-country flights between the East Coast Boston and NY markets, and the West Coast SF and LA markets in November 2023.
Market IDs and airports for the covered markets:
* NYC Metro, 31703: JFK, EWR, LGA
* SF Metro, 32457: SFO, OAK
* Boston Metro, 30721: BOS
* LA Metro, 32575: LAX, ONT, BUR, SNA


In [2]:
# Macros
flightdatapath = "../../shared/Datasets/Nov23_East_West_Flights.csv"
import pandas as pd
import numpy as np

In [3]:
nov23Flights = pd.read_csv(flightdatapath)
nov23Flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362 entries, 0 to 6361
Data columns (total 35 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Unnamed: 0          6362 non-null   int64  
 1   Year                6362 non-null   int64  
 2   Month               6362 non-null   int64  
 3   DayofMonth          6362 non-null   int64  
 4   DayOfWeek           6362 non-null   int64  
 5   FlightDate          6362 non-null   object 
 6   Operating_Airline   6362 non-null   object 
 7   Tail_Number         6360 non-null   object 
 8   Flight_Number       6362 non-null   int64  
 9   OriginAirportID     6362 non-null   int64  
 10  OriginCityMarketID  6362 non-null   int64  
 11  Origin              6362 non-null   object 
 12  OriginCityName      6362 non-null   object 
 13  OriginState         6362 non-null   object 
 14  DestAirportID       6362 non-null   int64  
 15  DestCityMarketID    6362 non-null   int64  
 16  Dest  

```<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6008 entries, 0 to 6007
Data columns (total 35 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Unnamed: 0          6008 non-null   int64  
 1   Year                6008 non-null   int64  
 2   Month               6008 non-null   int64  
 3   DayofMonth          6008 non-null   int64  
 4   DayOfWeek           6008 non-null   int64  
 5   FlightDate          6008 non-null   object 
 6   Operating_Airline   6008 non-null   object 
 7   Tail_Number         6006 non-null   object 
 8   Flight_Number       6008 non-null   int64  
 9   OriginAirportID     6008 non-null   int64  
 10  OriginCityMarketID  6007 non-null   float64
 11  Origin              6007 non-null   object 
 12  OriginCityName      6007 non-null   object 
 13  OriginState         6007 non-null   object 
 14  DestAirportID       6007 non-null   float64
 15  DestCityMarketID    6007 non-null   float64
 16  Dest                6007 non-null   object 
 17  DestCityName        6007 non-null   object 
 18  DestState           6007 non-null   object 
 19  CRSDepTime          6007 non-null   float64
 20  DepTime             5999 non-null   float64
 21  DepDelay            5999 non-null   float64
 22  CRSArrTime          6007 non-null   float64
 23  ArrTime             5998 non-null   float64
 24  ArrDelay            5985 non-null   float64
 25  Cancelled           6007 non-null   float64
 26  CancellationCode    8 non-null      object 
 27  Diverted            6007 non-null   float64
 28  CRSElapsedTime      6007 non-null   float64
 29  ActualElapsedTime   5985 non-null   float64
 30  AirTime             5985 non-null   float64
 31  Distance            6007 non-null   float64
 32  CarrierDelay        852 non-null    float64
 33  WeatherDelay        852 non-null    float64
 34  LateAircraftDelay   852 non-null    float64
dtypes: float64(18), int64(7), object(10)
memory usage: 1.6+ MB```

----

## Question 1 [24 points]
Write a function `getFlightDataSummary(flight_df)` that takes a DataFrame as input (the DataFrame is assumed to be formatted in the same way as `/shared/Datasets/Nov23_East_West_Flights.csv`).

The function should return a dictionary that computes the following:
* EastBoundFlights: the number of flights from West Coast Airports/Markets to East Coast Airports/Markets
* WestBoundFlights: the number of flights from East Coast Airports/Markets to West Coast Airports/Markets
* DepartureDelays: the number of flights that were delayed on departure (overall)
* ArrivalDelays: the number of flights that were delayed on arrival (overall)
* EastBoundDepartureDelays: the number of eastbound flights that were delayed on departure 
* EastBoundArrivalDelays: the number of eastbound flights that were delayed on arrival 
* WestBoundDepartureDelays: the number of westbound flights that were delayed on departure 
* WestBoundArrivalDelays: the number of westbound flights that were delayed on arrival 



In [4]:
from A2sub import getFlightDataSummary
import csv
import numpy as np
import pandas as pd

from typing import List, Dict, Tuple



def getFlightDataSummary(df: pd.DataFrame) -> Dict[str, int]:
    newdict = {'EastBoundFlights': 0, 'WestBoundFlights': 0, 'DepartureDelays': 0, 'ArrivalDelays': 0, 'EastBoundDepartureDelays': 0, 'EastBoundArrivalDelays': 0,
  'WestBoundDepartureDelays': 0,
  'WestBoundArrivalDelays': 0}
    bos_df = df[(df['Dest']=='BOS') & ((df['Origin']=='SFO') | (df['Origin']=='SNA') | (df['Origin']=='LAX') | (df['Origin']=='OAK') | (df['Origin']=='ONT') | (df['Origin']=='BUR'))]
    lgd_df = df[(df['Dest']=='LGA') & ((df['Origin']=='SFO') | (df['Origin']=='SNA') | (df['Origin']=='LAX') | (df['Origin']=='OAK') | (df['Origin']=='ONT') | (df['Origin']=='BUR'))]
    jfk_df = df[(df['Dest']=='JFK') & ((df['Origin']=='SFO') | (df['Origin']=='SNA') | (df['Origin']=='LAX') | (df['Origin']=='OAK') | (df['Origin']=='ONT') | (df['Origin']=='BUR'))]
    ewr_df = df[(df['Dest']=='EWR') & ((df['Origin']=='SFO') | (df['Origin']=='SNA') | (df['Origin']=='LAX') | (df['Origin']=='OAK') | (df['Origin']=='ONT') | (df['Origin']=='BUR'))]
    newdict['EastBoundFlights'] = len(bos_df) + len(lgd_df) + len(jfk_df) + len(ewr_df)
    oak_df = df[(df['Dest']=='OAK') & ((df['Origin']=='BOS') | (df['Origin']=='LGA') | (df['Origin']=='JFK') | (df['Origin']=='EWR'))]
    sfo_df = df[(df['Dest']=='SFO') & ((df['Origin']=='BOS') | (df['Origin']=='LGA') | (df['Origin']=='JFK') | (df['Origin']=='EWR'))]
    sna_df = df[(df['Dest']=='SNA') & ((df['Origin']=='BOS') | (df['Origin']=='LGA') | (df['Origin']=='JFK') | (df['Origin']=='EWR'))]
    bur_df = df[(df['Dest']=='BUR') & ((df['Origin']=='BOS') | (df['Origin']=='LGA') | (df['Origin']=='JFK') | (df['Origin']=='EWR'))]
    ont_df = df[(df['Dest']=='ONT') & ((df['Origin']=='BOS') | (df['Origin']=='LGA') | (df['Origin']=='JFK') | (df['Origin']=='EWR'))]
    lax_df = df[(df['Dest']=='LAX') & ((df['Origin']=='BOS') | (df['Origin']=='LGA') | (df['Origin']=='JFK') | (df['Origin']=='EWR'))]
    newdict['WestBoundFlights'] = len(oak_df) + len(sfo_df) + len(sna_df) + len(bur_df) + len(ont_df) + len(lax_df)
    dd = df[df['DepDelay'] > 0.0]
    ad = df[df['ArrDelay'] > 0.0]
    ebdd = len(bos_df[bos_df['DepDelay'] > 0.0]) + len(lgd_df[lgd_df['DepDelay'] > 0.0]) + len(jfk_df[jfk_df['DepDelay'] > 0.0]) + len(ewr_df[ewr_df['DepDelay'] > 0.0])
    ebad = len(bos_df[bos_df['ArrDelay'] > 0.0]) + len(lgd_df[lgd_df['ArrDelay'] > 0.0]) + len(jfk_df[jfk_df['ArrDelay'] > 0.0]) + len(ewr_df[ewr_df['ArrDelay'] > 0.0])
    wbdd = len(oak_df[oak_df['DepDelay'] > 0.0]) + len(sfo_df[sfo_df['DepDelay'] > 0.0]) + len(sna_df[sna_df['DepDelay'] > 0.0]) + len(bur_df[bur_df['DepDelay'] > 0.0]) + len(ont_df[ont_df['DepDelay'] > 0.0]) + len(lax_df[lax_df['DepDelay'] > 0.0])
    wbad = len(oak_df[oak_df['ArrDelay'] > 0.0]) + len(sfo_df[sfo_df['ArrDelay'] > 0.0]) + len(sna_df[sna_df['ArrDelay'] > 0.0]) + len(bur_df[bur_df['ArrDelay'] > 0.0]) + len(ont_df[ont_df['ArrDelay'] > 0.0]) + len(lax_df[lax_df['ArrDelay'] > 0.0])
    newdict['ArrivalDelays'] = len(ad)
    newdict['DepartureDelays'] = len(dd)
    newdict['EastBoundDepartureDelays'] = ebdd
    newdict['EastBoundArrivalDelays'] = ebad
    newdict['WestBoundDepartureDelays'] = wbdd
    newdict['WestBoundArrivalDelays'] = wbad
    return newdict
    

getFlightDataSummary(nov23Flights)  

{'EastBoundFlights': 3180,
 'WestBoundFlights': 3182,
 'DepartureDelays': 1945,
 'ArrivalDelays': 1695,
 'EastBoundDepartureDelays': 954,
 'EastBoundArrivalDelays': 795,
 'WestBoundDepartureDelays': 991,
 'WestBoundArrivalDelays': 900}

```
{'EastBoundFlights': 3180,
 'WestBoundFlights': 3182,
 'DepartureDelays': 1945,
 'ArrivalDelays': 1695,
 'EastBoundDepartureDelays': 954,
 'EastBoundArrivalDelays': 795,
 'WestBoundDepartureDelays': 991,
 'WestBoundArrivalDelays': 900}
```

---

## Question 2 [12 points]
Write a function `getDestinationSchedule(df, origin, destination)` that takes a DataFrame as input (the DataFrame is assumed to be formatted in the same way as /shared/Datasets/Nov23_East_West_Flights.csv), as well as two airport codes (three letters) as parameters. 

The function should return a DataFrame, grouped by the the following columns:

* Operating_Airline
* Type_of_Day (where type of day is 'weekday' or 'weekend', with 'weekend' days being Saturday and Sunday)

with the following value :

* Number_of_Flights: the number of flights for the corresponding (Airline,type of day) in the dataset.

Make sure to reset the index so that it goes from 0 to n-1

In [5]:
from A2sub import getDestinationSchedule

import csv
import numpy as np
import pandas as pd

from typing import List, Dict, Tuple

def getDestinationSchedule(df: pd.DataFrame, origin: str, destination: str) -> pd.DataFrame:
    origintodest = df[(df['Dest']==destination) & (df['Origin']==origin)]
    #print(origintodest.FlightDate.unique())
    #print(origintodest.Operating_Airline.unique())
    newestdict = []
    for i in origintodest.Operating_Airline.unique():
        newdf = origintodest[(origintodest['Operating_Airline']==i) & ((origintodest['DayOfWeek']==6) | (origintodest['DayOfWeek']==7))]
        newerdf = origintodest[(origintodest['Operating_Airline']==i) & (~((origintodest['DayOfWeek']==6) | (origintodest['DayOfWeek']==7)))]
        #print(i, len(newdf))
        #print(i, len(newerdf))
        newl = [i, "weekday", len(newerdf)]
        newl2 = [i, "weekend", len(newdf)]
        newestdict.append(newl)
        newestdict.append(newl2)
    newestdict.sort(key=lambda newestdict: newestdict[0])
    #print(newestdict)
    ds_df = pd.DataFrame(newestdict, columns=['Operating_Airline', 'Type_of_Day', 'Count_Flights'])
    return ds_df
    #print(len(origintodest))

getDestinationSchedule(nov23Flights,'SFO','BOS')

#getDestinationSchedule(nov23Flights,'JFK','OAK')

Operating_Airline Type_of_Day  Count_Flights
0                AS     weekday             21
1                AS     weekend              8
2                B6     weekday             98
3                B6     weekend             27
4                DL     weekday             42
5                DL     weekend             16
6                UA     weekday             98
7                UA     weekend             37

```
getDestinationSchedule(nov23Flights,'SFO','BOS')

	Operating_Airline	Type_of_Day	Count_Flights
    0	AS	weekday	21
    1	AS	weekend	8
    2	B6	weekday	98
    3	B6	weekend	27
    4	DL	weekday	42
    5	DL	weekend	16
    6	UA	weekday	98
    7	UA	weekend	37

getDestinationSchedule(nov23Flights,'BOS','LAX')

    	Operating_Airline	Type_of_Day	Count_Flights
    0	AA	weekday	43
    1	AA	weekend	16
    2	B6	weekday	91
    3	B6	weekend	27
    4	DL	weekday	69
    5	DL	weekend	27
    6	UA	weekday	43
    7	UA	weekend	16

getDestinationSchedule(nov23Flights,'JFK','SFO')

    	Operating_Airline	Type_of_Day	Count_Flights
    0	AA	weekday	84
    1	AA	weekend	28
    2	AS	weekday	103
    3	AS	weekend	40
    4	B6	weekday	125
    5	B6	weekend	36
    6	DL	weekday	147
    7	DL	weekend	47

getDestinationSchedule(nov23Flights,'JFK','OAK')

    Operating_Airline	Type_of_Day	Count_Flights

[Empty Dataframe]

```

----

## Question 3 [20 points]
Write a function `getAirCraftList(flight_df, metric='count')` that takes a DataFrame as input (the DataFrame is assumed to be formatted in the same way as /shared/Datasets/Nov23_East_West_Flights.csv).

The function should rank each aircraft based on the metric. The metric should be one of `'count','airtime','distance',delay`. If the metric entered is not in this list, your function should raise a `ValueError('Unknown metric')`. 
The ID of an aircraft is its `Tail_Number`

The function should return a dataframe that lists the top 10 Tail Numbers for the selected metric for November 23:
* count: number of flights, in new column *Count_Flights*
* airtime: total minutes in the air, in integer, in new column *Total_Air_Time*
* distance: total distance flown, in new column *Distance_Flown*
* delay: average delay of the aircraft on arrival (early arrivals should be counted as 0), rounded to two decimal places, in new column *Average_Delay*

Make sure to reset the index so that it goes from 0 to n-1

In [6]:
from A2sub import getAirCraftList

import csv
import numpy as np
import pandas as pd

def getAirCraftList(df: pd.DataFrame, metric: str='count') -> pd.DataFrame:
    #pass
    lisat = ['count', 'airtime', 'distance', 'delay']
    #print(df.Tail_Number.unique())
    try:
        ind = lisat.index(metric)
        if metric == 'count':
            lista = {}
            newlist = []
            for i in df.Tail_Number.unique():
                newdf = df[df['Tail_Number']==i]
                lista[i] = len(newdf)
            #print(lista)
            lista = dict(sorted(lista.items(), key=lambda item: item[1], reverse=True))
            #print(lista)
            first_ten = dict(list(lista.items())[:10])
            first_ten = first_ten.items()
            newerlist = []
            for i in first_ten:
                newerlist.append([i[0], i[1]])
            count_df = pd.DataFrame(newerlist, columns=['Tail_Number', 'Count_Flights'])
            return count_df
        else:
            if metric == 'airtime':
                newdict = {}
                for i in df.Tail_Number.unique():
                    newdf = df[df['Tail_Number']==i]
                    newdict[i] = newdf['AirTime'].sum()
                newdict = dict(sorted(newdict.items(), key=lambda item: item[1], reverse=True))
                first_ten_second = dict(list(newdict.items())[:10])
                first_ten_second = first_ten_second.items()
                newerlist_2 = []
                for i in first_ten_second:
                    newerlist_2.append([i[0], int(i[1])])
                at_df = pd.DataFrame(newerlist_2, columns=['Tail_Number', 'Total_Air_Time'])
                return at_df
            else:
                if metric == 'distance':
                    newdict = {}
                    for i in df.Tail_Number.unique():
                        newdf = df[df['Tail_Number']==i]
                        newdict[i] = newdf['Distance'].sum()
                    newdict = dict(sorted(newdict.items(), key=lambda item: item[1], reverse=True))
                    first_ten_second = dict(list(newdict.items())[:10])
                    first_ten_second = first_ten_second.items()
                    newerlist_2 = []
                    for i in first_ten_second:
                        newerlist_2.append([i[0], int(i[1])])
                    d_df = pd.DataFrame(newerlist_2, columns=['Tail_Number', 'Distance_Flown'])
                    return d_df
                else:
                    newdict = {}
                    nd = {}
                    for i in df.Tail_Number.unique():
                        newdf = df[df['Tail_Number']==i]
                        newerdf = newdf[newdf['ArrDelay'] > 0.0]
                        if(len(newerdf) > 0):
                            nd[i] = (newerdf['ArrDelay'].sum())/len(newdf)
                    nd = dict(sorted(nd.items(), key=lambda item: item[1], reverse=True))
                    fts = dict(list(nd.items())[:10])
                    fts = fts.items()
                    nl2 = []
                    for i in fts:
                        nl2.append([i[0], round(i[1], 2)])
                    delay_df = pd.DataFrame(nl2, columns=['Tail_Number', 'Average_Delay'])
                    return delay_df
    except ValueError as e:
        print('Unknown metric')
    
        
    
   

getAirCraftList(nov23Flights,'delay')
#getAirCraftList(nov23Flights)
#getAirCraftList(nov23Flights, 'distance')
#getAirCraftList(nov23Flights, 'taxis')

Tail_Number  Average_Delay
0      N198DN         287.00
1      N949JT         257.25
2      N569AS         202.00
3      N6708D         158.00
4      N394DL         148.50
5      N214AK         143.00
6      N629JB         141.00
7      N638JB         131.50
8      N677UA         126.50
9      N15710         126.00

```
getAirCraftList(nov23Flights,'airtime')

      	Tail_Number	Total_Air_Time
    0	N108NN	23012
    1	N109NN	22287
    2	N101NN	21896
    3	N712TW	21833
    4	N116AN	21591
    5	N105NN	21468
    6	N104NN	21250
    7	N102NN	20939
    8	N115NN	20590
    9	N103NN	19859

getAirCraftList(nov23Flights,'taxitime')

    ValueError: Unknown metric

getAirCraftList(nov23Flights,'distance')

     	Tail_Number	Distance_Flown
    0	N108NN	181976
    1	N109NN	178011
    2	N712TW	175848
    3	N101NN	173857
    4	N116AN	171505
    5	N105NN	171266
    6	N104NN	169049
    7	N102NN	165778
    8	N115NN	163697
    9	N721TW	157996

getAirCraftList(nov23Flights)
    	Tail_Number	Count_Flights
    0	N108NN	72
    1	N109NN	71
    2	N101NN	69
    3	N712TW	68
    4	N105NN	68
    5	N116AN	68
    6	N104NN	67
    7	N102NN	66
    8	N115NN	65
    9	N103NN	62

getAirCraftList(nov23Flights,'delay')

    	Tail_Number	Average_Delay
    0	N198DN	287.00
    1	N949JT	257.25
    2	N569AS	202.00
    3	N507JT	196.00
    4	N6708D	158.00
    5	N638JB	157.80
    6	N394DL	148.50
    7	N214AK	143.00
    8	N629JB	141.00
    9	N677UA	126.50
```

----

## Question 4 [20 points]
Write a function `getAirlineArrivalDelay(df, airlines = None,direction = None)` that takes a DataFrame as input (the DataFrame is assumed to be formatted in the same way as `/shared/Datasets/Nov23_East_West_Flights.csv`), and list of airline codes and a direction (East-West) as parameters. If the airlines are not specified, all airline are considered. If the direction is not specified, both directions are considered).

(Hint: you may have to create new columns)

The function should return a DataFrame, grouped by  the following columns
* Operating_Airline
* Direction

with the following values
* CountArrDelay: the number of flights for the corresponding (Airline,Direction) that *arrived* late 
* MeanArrDelay: the mean Arrival delay *in minutes, rounded to two decimal places) for the corresponding (Airline,Direction) that *arrived* late (so you need to ignore flights that were on time or early)

In [7]:
from A2sub import getAirlineDelay



def getAirlineDelay(df: pd.DataFrame, airlines: List[str]=None, direction: str=None) -> pd.DataFrame:
    #df.loc[(df['Dest'] == 'SFO' | df['Dest'] == 'OAK' | df['Dest'] == 'LAX') & (df['Origin'] == 'BOS' | df['Origin'] == 'JFK' | df['Origin'] == 'EWR' | df['Origin'] == 'LGA')  , 'Direction'] = 'East'
    #df.loc[(df['Dest'] == 'ONT' | df['Dest'] == 'BUR' | df['Dest'] == 'SNA') & (df['Origin'] == 'BOS' | df['Origin'] == 'JFK' | df['Origin'] == 'EWR' | df['Origin'] == 'LGA')  , 'Direction'] = 'East'
    df.loc[((df['Dest'] == 'SFO') | (df['Dest'] == 'OAK') | (df['Dest'] == 'LAX')) & ((df['Origin'] == 'BOS') | (df['Origin'] == 'JFK') | (df['Origin'] == 'EWR') | (df['Origin'] == 'LGA')) , 'Direction'] = 'West'
    df.loc[((df['Dest'] == 'ONT') | (df['Dest'] == 'BUR') | (df['Dest'] == 'SNA')) & ((df['Origin'] == 'BOS') | (df['Origin'] == 'JFK') | (df['Origin'] == 'EWR') | (df['Origin'] == 'LGA')) , 'Direction'] = 'West'
    df.loc[((df['Origin'] == 'SFO') | (df['Origin'] == 'OAK') | (df['Origin'] == 'LAX')) & ((df['Dest'] == 'BOS') | (df['Dest'] == 'JFK') | (df['Dest'] == 'EWR') | (df['Dest'] == 'LGA')) , 'Direction'] = 'East'
    df.loc[((df['Origin'] == 'ONT') | (df['Origin'] == 'BUR') | (df['Origin'] == 'SNA')) & ((df['Dest'] == 'BOS') | (df['Dest'] == 'JFK') | (df['Dest'] == 'EWR') | (df['Dest'] == 'LGA')) , 'Direction'] = 'East'
    listofall = []
    for i in df.Operating_Airline.unique():
        westidf = df[(df['Direction']=='West') & (df['Operating_Airline']==i)]
        eastidf = df[(df['Direction']=='East') & (df['Operating_Airline']==i)]
        westthatwerelate = westidf[westidf['ArrDelay'] > 0]
        eastlate = eastidf[eastidf['ArrDelay'] > 0]
        meandelayforwest = round(westthatwerelate['ArrDelay'].sum()/len(westthatwerelate), 2)
        eastlatedelay = round(eastlate['ArrDelay'].sum()/len(eastlate), 2)
        newline = [i, 'West', len(westthatwerelate), meandelayforwest]
        secondnewline = [i, 'East', len(eastlate), eastlatedelay]
        listofall.append(newline)
        listofall.append(secondnewline)
        listofall.sort(key=lambda listofall: (listofall[0], listofall[1]))
        q4_df = pd.DataFrame(listofall, columns=['Operating_Airline', 'Direction', 'CountArrDelay', 'MeanArrDelay'])
    if airlines == None and direction == None:
        return q4_df
    if airlines == None and direction != None:
        if direction == 'East':
            east_df = q4_df[q4_df['Direction'] == 'East']
            east_df.reset_index()
            return east_df
        else:
            west_df = q4_df[q4_df['Direction'] == 'West']
            west_df.reset_index()
            return west_df
    if direction == None and airlines != None:
        listofairlines = []
        for i in airlines:
            newdf = q4_df[q4_df['Operating_Airline']==i]
            listofairlines.append(newdf)
        c_df = pd.concat(listofairlines)
        c_df = c_df.sort_values(by=['Operating_Airline'], ascending=True)
        c_df = c_df.reset_index(drop=True)
        return c_df
    if direction != None and airlines != None:
        listofairlines = []
        for i in airlines:
            newdf = q4_df[q4_df['Operating_Airline']==i]
            listofairlines.append(newdf)
        c_df = pd.concat(listofairlines)
        c_df = c_df[c_df['Direction'] == direction]
        c_df = c_df.sort_values(by=['Operating_Airline'], ascending=True)
        c_df = c_df.reset_index(drop=True)
        return c_df
        
            
            
            
            

    
    #print(df.head())
getAirlineDelay(nov23Flights)
#getAirlineDelay(nov23Flights, direction='West') 
#getAirlineDelay(nov23Flights,['AA'],'West')
#getAirlineDelay(nov23Flights,['DL','AA'])

Operating_Airline Direction  CountArrDelay  MeanArrDelay
0                 AA      East            103         45.93
1                 AA      West            104         26.18
2                 AS      East             68         28.71
3                 AS      West            102         33.20
4                 B6      East            314         40.01
5                 B6      West            271         34.14
6                 DL      East            101         35.99
7                 DL      West            196         33.07
8                 NK      East             21         29.19
9                 NK      West             37         33.08
10                UA      East            188         29.96
11                UA      West            190         34.95

```
getAirlineDelay(nov23Flights)

	Operating_Airline	Direction	CountArrDelay	MeanArrDelay
    0	AA	East	103	45.93
    1	AA	West	104	26.18
    2	AS	East	68	28.71
    3	AS	West	102	33.20
    4	B6	East	314	40.01
    5	B6	West	271	34.14
    6	DL	East	101	35.99
    7	DL	West	196	33.07
    8	NK	East	21	29.19
    9	NK	West	37	33.08
    10	UA	East	188	29.96
    11	UA	West	190	34.95

getAirlineDelay(nov23Flights,['AA'],'West')

	Operating_Airline	Direction	CountArrDelay	MeanArrDelay
    0	AA	West	104	26.18

getAirlineDelay(nov23Flights,['AA','DL'])

    Operating_Airline	Direction	CountArrDelay	MeanArrDelay
    0	AA	East	103	45.93
    1	AA	West	104	26.18
    2	DL	East	101	35.99
    3	DL	West	196	33.07

getAirlineDelay(nov23Flights,direction='East')

	Operating_Airline	Direction	CountArrDelay	MeanArrDelay
    0	AA	East	103	45.93
    1	AS	East	68	28.71
    2	B6	East	314	40.01
    3	DL	East	101	35.99
    4	NK	East	21	29.19
    5	UA	East	188	29.96
```

---

## Question 5 [24 points]
Write a function `getPercentageLate(df, airports = None)` that takes a DataFrame as input (the DataFrame is assumed to be formatted in the same way as `/shared/Datasets/Nov23_East_West_Flights.csv`) and a list of airports codes. If the airports are not specified, all airports are considered.

(Hint: you may have to create new columns and helper functions)

The function should return a DataFrame, grouped by the the following columns
* Airport
* Time of planned Departure as `TimeOfDay`, where the time of Departure (CRSDepTime) is grouped as (inclusive,exclusive):
    * Early Morning: 2am to 6am
    * Morning: 6am to 10am
    * Midday: 10am to 2pm
    * Afternoon: 2pm to 6pm
    * Evening: 6pm to 10pm
    * Late Night: 10pm to 2am  

with the following value
* NumberLateDeparture: the number of flights that left late in the group
* TotalDeparture: the total number of flights that left in the group
* PercentageLateDeparture: the percentage of flights that left late in the group (rounded to 2 decimal)

Your result DataFrame should be ordered by TimeOfDay, not alphabetical, but from Early Morning to Late Night. (hint, use pd.Categorical as a type for TimeOfDay)


In [119]:
from A2sub import getPercentageLate

import csv
import numpy as np
import pandas as pd

from typing import List, Dict, Tuple

def getPercentageLate(df: pd.DataFrame, airports: List[str]= None) -> pd.DataFrame:
    df.loc[((df['CRSDepTime'] >= 200.0) & (df['CRSDepTime'] < 600.0)), 'TimeOfDay'] = 'Early Morning'
    df.loc[((df['CRSDepTime'] >= 600.0) & (df['CRSDepTime'] < 1000.0)), 'TimeOfDay'] = 'Morning'
    df.loc[((df['CRSDepTime'] >= 1000.0) & (df['CRSDepTime'] < 1400.0)), 'TimeOfDay'] = 'Midday'
    df.loc[((df['CRSDepTime'] >= 1400.0) & (df['CRSDepTime'] < 1800.0)), 'TimeOfDay'] = 'Afternoon'
    df.loc[((df['CRSDepTime'] >= 1800.0) & (df['CRSDepTime'] < 2200.0)), 'TimeOfDay'] = 'Evening'
    df.loc[((df['CRSDepTime'] >= 2200.0) | (df['CRSDepTime'] < 200.0)), 'TimeOfDay'] = 'Late Night'
    totallist = []
    for i in df.Origin.unique():
        em_df = df[(df['Origin']==i) & (df['TimeOfDay']=='Early Morning')]
        emdf = em_df[em_df['DepDelay'] > 0.0]
        if len(emdf) != 0 and len(em_df) != 0:
            emlist = [1, i, len(emdf), len(em_df), round((len(emdf)/len(em_df))*100, 2)]
            totallist.append(emlist)
        m_df = df[(df['Origin']==i) & (df['TimeOfDay']=='Morning')]
        mdf = m_df[m_df['DepDelay'] > 0.0]
        if len(mdf) != 0 and len(m_df) != 0:
            mlist = [2, i, len(mdf), len(m_df), round((len(mdf)/len(m_df))*100, 2)]
            totallist.append(mlist)
        mid_df = df[(df['Origin']==i) & (df['TimeOfDay']=='Midday')]
        middf = mid_df[mid_df['DepDelay'] > 0.0]
        if len(middf) != 0 and len(mid_df) != 0:
            midlist = [3, i, len(middf), len(mid_df), round((len(middf)/len(mid_df))*100, 2)]
            totallist.append(midlist)
        a_df = df[(df['Origin']==i) & (df['TimeOfDay']=='Afternoon')]
        adf = a_df[a_df['DepDelay'] > 0.0]
        if len(adf) != 0 and len(a_df) != 0:
            alist = [4, i, len(adf), len(a_df), round((len(adf)/len(a_df))*100, 2)]
            totallist.append(alist)
        e_df = df[(df['Origin']==i) & (df['TimeOfDay']=='Evening')]
        edf = e_df[e_df['DepDelay'] > 0.0]
        if len(edf) != 0 and len(e_df) != 0:
            elist = [5, i, len(edf), len(e_df), round((len(edf)/len(e_df))*100, 2)]
            totallist.append(elist)
        ln_df = df[(df['Origin']==i) & (df['TimeOfDay']=='Late Night')]
        lndf = ln_df[ln_df['DepDelay'] > 0.0]
        if len(lndf) != 0 and len(ln_df) != 0:
            lnlist = [6, i, len(lndf), len(ln_df), round((len(lndf)/len(ln_df))*100, 2)]
            totallist.append(lnlist)
    totallist = sorted(totallist, key=lambda x: [x[0], x[1]])
    entire_Df = pd.DataFrame(totallist, columns=["TimeOfDay", "Origin", "NumberLateDeparture", "TotalDeparture", "PercentageLateDeparture"])
    entire_Df['TimeOfDay'] = entire_Df['TimeOfDay'].replace(1, 'Early Morning')
    entire_Df['TimeOfDay'] = entire_Df['TimeOfDay'].replace(2, 'Morning')
    entire_Df['TimeOfDay'] = entire_Df['TimeOfDay'].replace(3, 'Midday')
    entire_Df['TimeOfDay'] = entire_Df['TimeOfDay'].replace(4, 'Afternoon')
    entire_Df['TimeOfDay'] = entire_Df['TimeOfDay'].replace(5, 'Evening')
    entire_Df['TimeOfDay'] = entire_Df['TimeOfDay'].replace(6, 'Late Night')
    #print(entire_Df.head())
    if airports == None:
        return entire_Df
    else:
        listofairlines = []
        for i in airports:
                newdf = entire_Df[entire_Df['Origin']==i]
                newdf.loc[ newdf['TimeOfDay'] == 'Early Morning', 'TimeOfDay'] = 1
                newdf.loc[ newdf['TimeOfDay'] == 'Morning', 'TimeOfDay'] = 2
                newdf.loc[ newdf['TimeOfDay'] == 'Midday', 'TimeOfDay'] = 3
                newdf.loc[ newdf['TimeOfDay'] == 'Afternoon', 'TimeOfDay'] = 4
                newdf.loc[ newdf['TimeOfDay'] == 'Evening', 'TimeOfDay'] = 5
                newdf.loc[ newdf['TimeOfDay'] == 'Late Night', 'TimeOfDay'] = 6
                listofairlines.append(newdf)
        c_df = pd.concat(listofairlines)
        c_df = c_df.sort_values(by='TimeOfDay')
        c_df.loc[ c_df['TimeOfDay'] == 1, 'TimeOfDay'] = 'Early Morning'
        c_df.loc[ c_df['TimeOfDay'] == 2, 'TimeOfDay'] = 'Morning'
        c_df.loc[ c_df['TimeOfDay'] == 3, 'TimeOfDay'] = 'Midday'
        c_df.loc[ c_df['TimeOfDay'] == 4, 'TimeOfDay'] = 'Afternoon'
        c_df.loc[ c_df['TimeOfDay'] == 5, 'TimeOfDay'] = 'Evening'
        c_df.loc[ c_df['TimeOfDay'] == 6, 'TimeOfDay'] = 'Late Night'
        c_df = c_df.reset_index(drop=True)
        return c_df
#getPercentageLate(nov23Flights)
getPercentageLate(nov23Flights, ['JFK'])

TimeOfDay Origin  NumberLateDeparture  TotalDeparture  \
0  Early Morning    JFK                    1               8   
1        Morning    JFK                   84             466   
2         Midday    JFK                   39             203   
3      Afternoon    JFK                  123             400   
4        Evening    JFK                  148             407   
5     Late Night    JFK                   21              43   

   PercentageLateDeparture  
0                    12.50  
1                    18.03  
2                    19.21  
3                    30.75  
4                    36.36  
5                    48.84

```
getPercentageLate(nov23Flights,['JFK'])

     	TimeOfDay	Origin	NumberLateDeparture	TotalDeparture	PercentageLateDeparture
    0	Early Morning	JFK	1	8	12.50
    1	Morning	JFK	84	466	18.03
    2	Midday	JFK	39	201	19.40
    3	Afternoon	JFK	123	400	30.75
    4	Evening	JFK	148	406	36.45
    5	Late Night	JFK	21	43	48.84

getPercentageLate(nov23Flights)

     	TimeOfDay	Origin	NumberLateDeparture	TotalDeparture	PercentageLateDeparture
    0	Early Morning	JFK	1	8	12.50
    1	Early Morning	LAX	5	29	17.24
    2	Morning	SNA	3	30	10.00
    3	Morning	LGA	1	4	25.00
    4	Morning	LAX	80	489	16.36
    5	Morning	JFK	84	466	18.03
    6	Morning	EWR	73	261	27.97
    7	Morning	BOS	97	291	33.33
    8	Morning	SFO	61	332	18.37
    9	Midday	SNA	21	59	35.59
    10	Midday	SFO	146	391	37.34
    11	Midday	LAX	159	458	34.72
    12	Midday	JFK	39	201	19.40
    13	Midday	EWR	47	171	27.49
    14	Midday	BOS	30	92	32.61
    15	Afternoon	BOS	42	125	33.60
    16	Afternoon	EWR	89	254	35.04
    17	Afternoon	JFK	123	400	30.75
    18	Afternoon	LAX	67	221	30.32
    19	Afternoon	SFO	60	171	35.09
    20	Evening	SNA	1	29	3.45
    21	Evening	BOS	62	160	38.75
    22	Evening	BUR	20	30	66.67
    23	Evening	EWR	128	286	44.76
    24	Evening	JFK	148	406	36.45
    25	Evening	LAX	80	196	40.82
    26	Evening	OAK	8	19	42.11
    27	Evening	SFO	21	73	28.77
    28	Late Night	SFO	102	335	30.45
    29	Late Night	ONT	11	30	36.67
    30	Late Night	OAK	3	6	50.00
    31	Late Night	LAX	106	279	37.99
    32	Late Night	BOS	6	9	66.67
    33	Late Night	JFK	21	43	48.84

---